blockmanager.go

```diff
@@ -319,6 +324,492 @@ func (b *blockManager) handleDonePeerMsg(peers *list.List, sp *ServerPeer) {
+// map of responses.
+func (b *blockManager) getCFHeaders(height uint32,
+	fType wire.FilterType) map[*ServerPeer]*wire.MsgCFHeaders {
+	headers := make(map[*ServerPeer]*wire.MsgCFHeaders)
+	stopHeader, err := b.server.BlockHeaders.FetchHeaderByHeight(
+		height + wire.MaxCFHeadersPerMsg - 1)
+	if err != nil {
+		return headers
+	}
+	stopHash := stopHeader.BlockHash()
+	msg := wire.NewMsgGetCFHeaders(fType, height, &stopHash)
+	b.server.queryAllPeers(
+		msg,
+		func(sp *ServerPeer, resp wire.Message, quit chan<- struct{},
+			peerQuit chan<- struct{}) {
+			switch m := resp.(type) {
+			case *wire.MsgCFHeaders:
+				if m.StopHash == stopHash &&
+					m.FilterType == fType {
+					headers[sp] = m
+					close(peerQuit)
+				}
+			}
+		},
+	)
+	return headers
+}
+
+// getCheckpts runs a query for cfcheckpts against all peers and returns a map
+// of responses.
+func (b *blockManager) getCheckpts(lastHash *chainhash.Hash,
+	fType wire.FilterType) map[*ServerPeer][]*chainhash.Hash {
+
+	checkpoints := make(map[*ServerPeer][]*chainhash.Hash)
+	getCheckptMsg := wire.NewMsgGetCFCheckpt(fType, lastHash)
+	b.server.queryAllPeers(
+		getCheckptMsg,
+		func(sp *ServerPeer, resp wire.Message, quit chan<- struct{},
+			peerQuit chan<- struct{}) {
+			switch m := resp.(type) {
+			case *wire.MsgCFCheckpt:
+				if m.FilterType == fType &&
+					m.StopHash == *lastHash {
+					checkpoints[sp] = m.FilterHeaders
+					close(peerQuit)
+				}
+			}
+		},
+	)
+	return checkpoints
+}
+
+// checkCFCheckptSanity checks whether all peers which have responded agree.
+// If so, it returns -1; otherwise, it returns the earliest index at which at
+// least one of the peers differs. The checkpoints are also checked against the
+// existing store up to the tip of the store. If all of the peers match but
+// the store doesn't, the height at which the mismatch occurs is returned.
+func checkCFCheckptSanity(cp map[*ServerPeer][]*chainhash.Hash,
+	headerStore *headerfs.FilterHeaderStore) (int, error) {
+
+	// Get the known best header to compare against checkpoints.
+	_, storeTip, err := headerStore.ChainTip()
+	if err != nil {
+		return 0, err
+	}
+
+	// Determine the maximum length of each peer's checkpoint list. If they
+	// differ, we don't return yet because we want to make sure they match
+	// up to the shortest one.
+	maxLen := 0
+	for _, checkpoints := range cp {
+		if len(checkpoints) > maxLen {
+			maxLen = len(checkpoints)
+		}
+	}
+
+	// Compare the actual checkpoints against each other and anything
+	// stored in the header store.
+	for i := 0; i < maxLen; i++ {
+		var checkpoint chainhash.Hash
+		for _, checkpoints := range cp {
+			if i >= len(checkpoints) {
+				return i, nil
+			}
+			if checkpoint == (chainhash.Hash{}) {
+				checkpoint = *checkpoints[i]
+			}
+			if checkpoint != *checkpoints[i] {
+				return i, nil
+			}
+		}
+		if uint32((i+1)*wire.CFCheckptInterval) <= storeTip {
+			header, err := headerStore.FetchHeaderByHeight(
```

Comment:
Right, this code is run on every start up to catch up any checkpoints that have happened since we were offline. In addition, this plays into the TODO above for counteracting eclipse attacks after restarting and getting honest peers.